---
title: "Ch 1.1: Alpha Diversity Analyses of Age and Maturation"
author: "Mauna Dasari"
date: "Oct 2, 2018"
output: 
  html_document:
    toc: true
    toc_depth: 6
    toc_float: true
---

```{r setup, include=FALSE, eval=T}
knitr::opts_chunk$set(echo = F, warning=F,message=F,cache=2,
                      fig.width=30,fig.height=20)

library(tidyverse); library(kableExtra);library(knitr);
library(vegan);library(pander);library(dplyr)

```

This file will focus on processing the raw data file into clean data for the machine learning algorithms. The code for calculating alpha diversity is also in this file. Because the files are rather large, I've left in (but commented out) lines where an intermediate file was saved for quicker processing at later points. 

# Taxonomy Cleaning
```{r clean_taxonomy}
#silva_raw<-readRDS("Ch1/raw_data/IDTAXA_silva_assignment.rds")
silva_raw<-readRDS("../data/IDTAXA_silva_assignment.rds")

dim(silva_raw) #45350x7, the rows are seq and the col are taxa designations
silva<-as_tibble(silva_raw,rownames=NA) #convert it from a matrix to a tibble, preserve the rownames
silva$ASV<-rownames(silva)
#What's in the silva file?
silva$taxa_phyla<-paste(silva$domain,silva$phylum,sep="_")
silva$taxa_phyla<-gsub("-",".",silva$taxa_phyla)

silva$taxa_order<-paste(silva$domain,silva$phylum,silva$order, sep="_")
silva$taxa_order<-gsub("-",".",silva$taxa_order)

silva$taxa_family<-paste(silva$domain,silva$phylum,silva$order,silva$family, sep="_")
silva$taxa_family<-gsub("-",".",silva$taxa_family)

silva$taxa_genus<-paste(silva$domain,silva$phylum,silva$order,silva$family,silva$genus, sep="_")
silva$taxa_genus<-gsub("-",".",silva$taxa_genus)

silva$ASV_id<-paste("ASV",1:nrow(silva),sep=" ")
#adding ASV_id to simplify naming schemes 

taxonomy<-silva%>%
  select(contains("taxa_"))
taxonomy$ASVs<-rownames(silva)
#saveRDS(taxonomy,"../data/taxonomy_assignment_clean.rds")
```

# Removing Singletons and Reads Mapped to Chloroplasts or Mitochondria
```{r data_wrangling.import_reads}
#Load the data!
reads_raw<-readRDS("../in/merged_baboon_data_with_metadata_preset_more_blanks_greater_1000.rds")

pres_abs<-reads_raw
pres_abs[pres_abs>0]<-1 #ASV as the colnames, sampleID as rownames
#saveRDS(pres_abs, "../out/presence_absence_raw.RDS")

#pres_abs<-readRDS("../data/presence_absence_raw.RDS")
pres_abs_t<-as.data.frame(t(pres_abs)) #transpose, samples are columns and rows are ASVs. 
pres_abs_t$ASVSum<-rowSums(pres_abs_t) #sum across the rows to find out if that ASV is a singleton or not.
pres_abs_t$ASVs<-rownames(pres_abs_t) #save the ASV names!
pres_abs_t2<-pres_abs_t%>%
    filter(ASVSum>2)#filter by singleton or doubleton status 
#saveRDS(pres_abs_t2, "../out/presence_absence_no_singletons_ASVSum_incl.RDS")
#pres_abs_t2<-readRDS("../out/presence_absence_no_singletons_ASVSum_incl.RDS")

#pres_abs<-readRDS("../data/presence_absence_no_singletons_ASVSum_incl.RDS")
#taxonomy<-readRDS("../data/taxonomy_assignment_clean.rds")
reads_taxonomy<-merge(pres_abs, taxonomy, by="ASVs")

reads_taxonomy_no_chl_mt<-reads_taxonomy %>% 
  filter(taxa_order!="Bacteria_Cyanobacteria_Chloroplast") %>% 
  filter(taxa_family!="Bacteria_Proteobacteria_Rickettsiales_Mitochondria") #switches from 19182 to 19030, filtering out 152 ASVs
#saveRDS(reads_taxonomy_no_chl_mt,"../data/reads_taxonomy_no_chl_mt.rds")
#reads_taxonomy_no_chl_mt<-readRDS("../data/reads_taxonomy_no_chl_mt.rds")

#filter our abundance table
readst<-as.data.frame(t(reads_raw))#transpose, samples are columns and rows are ASVs. 

readst$ASVs<-rownames(readst) #save the ASV names!
readst2<-readst%>%
    filter(ASVs %in% reads_taxonomy_no_chl_mt$ASVs) #filter the larger table by the smaller table, previously filtered by pres_abs_t2
rownames(readst2)<-readst2$ASVs
readst2<-readst2 %>% dplyr::select(c(-ASVs)) 

reads_final<-as.data.frame(t(readst2))

reads_final$DADA_id<-rownames(reads_final) #make the sample IDs accessible
#write_rds(reads_final, "../data/reads_intermediate_no_chl_mt.rds")
#reads3<-readRDS("../out/reads_intermediate_no_chl_mt.rds")

metadata<-readRDS("../data/metadata_full.rds")
monster_data<-merge(metadata, reads_final, by="DADA_id") 
```

# Calculating Alpha Diversity Metrics
```{r alpha_diversity}
monster_data2<-monster_data%>%dplyr::select(-c("sid","did","tid","DADA_id","plate","sname","extract_dna_conc_ng", "loading_plate_location","collection_date","sex","birth","bstatus","matured","matgrp"))%>%
    map_if(is.factor,as.character)%>%
    map_if(is.character,as.numeric)%>%
    as_data_frame

monster_data$read_count<-rowSums(monster_data2)

rsummary<-data.frame(DADA_id = monster_data$DADA_id,
                    read_count = monster_data$read_count)

#richness outputs a matrix
Rrichness<-estimateR(monster_data2) 
obs.richness<-Rrichness["S.obs",]
rsummary$richness<-obs.richness

#calculate Shannon's H
rsummary$ShannonH<-diversity(monster_data2,index="shannon")
beep(sound=5)
#And Simpson's, 1 is infinite diversity and 0 is no diversity
rsummary$Simpson<-diversity(monster_data2,index="simpson")
beep(sound=5)

Hill<-renyi(monster_data2, hill=T)*1 
rsummary$Hill.q1<-Hill$`1`
rsummary$Hill.q2<-Hill$`2`
#saveRDS(rsummary,"../out/reads_summary2_no_singletons_psuedopooled_bstatus6mo_no_chl_mt.rds") 
```

The original dataset was comprised of 45,350 Amplicon Sequence Variants (ASVs) over 13,563 samples. After filtering for singletons, we have 16,258 ASVs over 13,563 samples. The highest number of reads in a sample was `r max(rsummary$total_reads)` and the lowest was `r min(rsummary$total_reads)` with a median read count of `r median(rsummary$total_reads)`. 
